os/drivers/usbhost/uvc/usbhost_uvc_ctrl.c

```diff
@@ -0,0 +1,1357 @@
+ *
+ * Returned Values:
+ *   OK on success, appropriate error code on failures
+ *
+ ****************************************************************************/
+static int allocate_ctrl_vals(struct uvc_control *ctrl)
+{
+	int ret = OK;
+	/* Allocate memory for each of the controls */
+	ctrl->min = kmm_zalloc(ctrl->info.size);
+	if (ctrl->min == NULL) {
+		ret = -ENOMEM;
+		goto done;
+	}
+
+	ctrl->max = kmm_zalloc(ctrl->info.size);
+	if (ctrl->max == NULL) {
+		ret = -ENOMEM;
+		goto done;
+	}
+
+	ctrl->cur = kmm_zalloc(ctrl->info.size);
+	if (ctrl->cur == NULL) {
+		ret = -ENOMEM;
+		goto done;
+	}
+
+	ctrl->step = kmm_zalloc(ctrl->info.size);
+	if (ctrl->step == NULL) {
+		ret = -ENOMEM;
+		goto done;
+	}
+
+	ctrl->def = kmm_zalloc(ctrl->info.size);
+	if (ctrl->def == NULL) {
+		ret = -ENOMEM;
+		goto done;
+	}
+
+	ctrl->bak = kmm_zalloc(ctrl->info.size);
+	if (ctrl->bak == NULL) {
+		ret = -ENOMEM;
+		goto done;
+	}
+done:
+	if (ret < 0) {
+		if (ctrl->min) {
+			kmm_free(ctrl->min);
+		}
+		if (ctrl->max) {
+			kmm_free(ctrl->max);
+		}
+		if (ctrl->cur) {
+			kmm_free(ctrl->cur);
+		}
+		if (ctrl->bak) {
+			kmm_free(ctrl->bak);
+		}
+		if (ctrl->step) {
+			kmm_free(ctrl->step);
+		}
+		if (ctrl->def) {
+			kmm_free(ctrl->def);
+		}
+	}
+	return ret;
+}
+
+/****************************************************************************
+ * Name: uvc_ctrl_add_info
+ *
+ * Description:
+ *   This function allocate memory for min/max/default/current/backup, retrives
+ *   the capabilities supported by the device.
+ *
+ *   This function is called from the uvc_init_device_ctrl method.
+ *
+ * Input Parameters:
+ *   dev - reference to driver object.
+ *   ctrl - reference to the conrol to be initialized.
+ *   info - reference to the control info.
+ *
+ * Returned Values:
+ *   OK on success, appropriate error code on failures
+ *
+ ****************************************************************************/
+int uvc_ctrl_add_info(uvc_state_t *dev, struct uvc_control *ctrl, const struct uvc_control_info *info)
+{
+	uint8_t *data;
+	int ret = 0;
+
+	ctrl->info = *info;
+	INIT_QUEUE_HEAD(&ctrl->info.mappings);
+
+	/* Allocate an array to save control values (cur, def, max, etc.) */
+	ret = allocate_ctrl_vals(ctrl);
+	if (ret < 0) {
+		udbg("Failed to allocate memory!!\n");
+		goto done;
+	}
+
+	/* Ignore the error and work with default values */
+	data = kmm_zalloc(1);
+	if (data == NULL) {
+		udbg("Failed to allocate memory!!\n");
+		ret = -ENOMEM;
+		goto done;
+	}
+
+	ret = uvc_query_ctrl(dev, UVC_GET_INF, ctrl->entity->id, dev->intfnum, ctrl->info.selector, data, 1);
+	if (ret >= 0) {
+		ctrl->info.flags |= (data[0] & UVC_CONTROL_CAP_GET ? F_GET_CUR : 0)
+							| (data[0] & UVC_CONTROL_CAP_SET ? F_SET_CUR : 0)
+							| (data[0] & UVC_CONTROL_CAP_AUTOUPDATE ? F_AUTO : 0)
+							| (data[0] & UVC_CONTROL_CAP_ASYNCHRONOUS ? F_ASYNCH : 0);
+		ret = OK;
+	}
+
+	kmm_free(data);
+
+	CTRL_SET_SUPPORTED(ctrl->status);
+
+done:
```

Comment:
If the label only returns, the "goto" could be avoided?

Reply:
I thought better to have one return instead!

Reply:
Would avoiding a goto label be better?

Reply:
Any benefit with avoiding goto labeling?

Reply:
Goto can cause 'spaghetti code', i.e. code becomes difficult to trace.
Though it might not happen here but due to that, generally considered a good practice to avoid goto statements.

Reply:
Corrected!